docs/adr/0004-instrumentation.md

```diff
@@ -0,0 +1,99 @@
+
+- Status: proposed
+- Deciders: Christina Harlow, Linh Nguyen, Rob Miller
+- Date: [2022-04-05
+
+Technical Story: https://github.com/mozilla/glam/issues/1847
+
+## Context and Problem Statement
+
+In order to help us with various product decisions and overall product quality,
+we want to understand how GLAM is normally used: Where does it fit in a typical
+user's workflow (a starting point, an endpoint, or something in between)?
+
+We can accomplish this by gathering data such as how a typical user session
+looks like, most and least looked at probes, how probe browsing behaves over
+short and long periods of time and whether it correlates with other events such
+as certain product releases.
+
+Additionally, as a more time-sensitive issue, GLAM ETL does a lot of work and
+takes over 12 hours to complete. We intend to reduce GLAM ETL time by finding
+the probes that are most likely to be requested by users from the aforementioned
+data and have the ETL build only those probes. This document also presents a
+plan for ad hoc probe building, which needs to happen when someone requests a
+probe that had not been pre-built by the ETL.
+
+## Decision Drivers
+
+- Usage metrics collection must be reliable and cannot be bypassed or influenced
+  by other systems such as browsers' adblocks
+- GLAM ETL should consume this service so it can optimize its work, only
+  processing probes that are likely to be looked at
+- GLAM App should consume this service so it can trigger an ad hoc calculation
+  whenever a probe that was not processed by GLAM ETL is requested
+- Ad hoc probe building should be independent of GLAM App because it's likely to
+  be CPU/memory intensive, with different needs in terms of scalability
+
+## Considered Options
+
+- 1: Collect metrics on the front end (via a library such as Open Telemetry
+  Instrumentation or a custom-built solution) and send them to the GLAM back end
+  server to be stored in a table in the GLAM App Postgres DB. Expose meaningful
+  api endpoints for usaga data consumption by GLAM App and and GLAM ETL. This
+  provides flexibility to adapt to future requirements and "immunity" to
+  adblocks, but might be longer to implement
+- 2: Use Gleanjs, a JS library for telemetry, which would be quicker than a
+  custom implementation. It's also good because of dogfooding and internal
+  support. The downside here is it can be bypassed by adblockers, because it
+  sends data to a different endpoint
+- 3: Google analytics: currently available, it's quick and easy to implement,
+  but doesn't give us the insight into e.g. session-scoped data and can also be
+  bypassed by adblockers
+- 4: Option #1, but instead of using GLAM's Postgres DB, save metrics to a
+  BigQuery table. This would make it easy for the ETL to consume the data, but
+  will most likely slow down some GLAM App operations due to BigQuery's
+  slower-than-OLTP response times
```

Comment:
How are you planning to consume the data? I suspect that, once you have some recorded metrics, you'd likely need an ETL job or something like that to summarize them. At that point, you really need the _output_ of the ETL to be in Postgres DB, not the raw data, so the latter could live in either places. What do you think?

Reply:
I think this answer above might have confused @cmharlow. It looks like storing data on BQ once we start collecting glam-wide metrics is the way to go. But as far as recording usage on probe requests only, we'll store it in glam's psql db

Reply:
It sounds like the design decision here is to add a GLAM API endpoint, which triggers a downstream postgres query to return a list of recently used probes. This will be used by the ETL system  to determine a subset of all probes to load back into GLAM's backend postgres db.

In general we should try to decouple the frontend application's health as a requirement to loading it's data. How will ETL processes behave when GLAM is down? LRU cache/value? What if there is an issue with generating a valid API and/or postgres response?